commons-math/src/main/java/org/apache/commons/math4/optim/univariate/BracketFinder.java

```diff
@@ -0,0 +1,314 @@
+                    xB = xC;
+                    xC = w;
+                    w = xC + GOLD * (xC - xB);
+                    fB = fC;
+                    fC =fW;
+                    fW = eval.value(w);
+                }
+            } else {
+                w = xC + GOLD * (xC - xB);
+                fW = eval.value(w);
+            }
+
+            xA = xB;
+            fA = fB;
+            xB = xC;
+            fB = fC;
+            xC = w;
+            fC = fW;
+        }
+
+        lo = xA;
+        fLo = fA;
+        mid = xB;
+        fMid = fB;
+        hi = xC;
+        fHi = fC;
+
+        if (lo > hi) {
+            double tmp = lo;
+            lo = hi;
+            hi = tmp;
+
+            tmp = fLo;
+            fLo = fHi;
+            fHi = tmp;
+        }
+    }
+
+    /**
+     * @return the number of evalutations.
+     */
+    public int getMaxEvaluations() {
+        return maxEvaluations;
+    }
+
+    /**
+     * @return the number of evalutations.
+     */
+    public int getEvaluations() {
+        return evaluations;
+    }
+
+    /**
+     * @return the lower bound of the bracket.
+     * @see #getFLo()
+     */
+    public double getLo() {
+        return lo;
+    }
+
+    /**
+     * Get function value at {@link #getLo()}.
+     * @return function value at {@link #getLo()}
+     */
+    public double getFLo() {
+        return fLo;
+    }
+
+    /**
+     * @return the higher bound of the bracket.
+     * @see #getFHi()
+     */
+    public double getHi() {
+        return hi;
+    }
+
+    /**
+     * Get function value at {@link #getHi()}.
+     * @return function value at {@link #getHi()}
+     */
+    public double getFHi() {
+        return fHi;
+    }
+
+    /**
+     * @return a point in the middle of the bracket.
+     * @see #getFMid()
+     */
+    public double getMid() {
+        return mid;
+    }
+
+    /**
+     * Get function value at {@link #getMid()}.
+     * @return function value at {@link #getMid()}
+     */
+    public double getFMid() {
+        return fMid;
+    }
+
+    /**
+     * Utility for incrementing a counter at each function evaluation.
+     */
+    private class FunctionEvaluator {
+        /** Function. */
+        private final UnivariateFunction func;
+        /** Counter. */
+        private final Incrementor inc;
+
+        /**
+         * @param func Function.
+         */
+        FunctionEvaluator(UnivariateFunction func) {
+            this.func = func;
+            inc = Incrementor.create().withMaximalCount(maxEvaluations);
+            evaluations = 0;
+        }
+
+        /**
+         * @param x Argument.
+         * @return {@code f(x)}
+         * @throws TooManyEvaluationsException if the maximal number of evaluations is
+         * exceeded.
+         */
+        double value(double x) {
+            try {
+                inc.increment();
+                evaluations = inc.getCount();
+            } catch (MaxCountExceededException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)